### PR TITLE
port cuda solver without triangular solver to hip

### PR DESCRIPTION
**Pull request by yhmtsai**
This PR ports cuda solver without triangular solver to hip

The test of preconditioner needs to use `ginkgo_create_test` not `ginkgo_create_hip_test`.
Otherwise, the stop factory is always empty.

The triangular solver needs some extra works, so I do not put it in this PR.

Note: Some conversion fails from cuda to hip.
The block_size or grid_size is too long to hipify-perl can not recognize it correctly, so need some fixes by hand.
[fail1-1](https://github.com/ginkgo-project/ginkgo/blob/78d2113dca67afea0d5567dfd511b89bb083e2b9/cuda/solver/gmres_kernels.cu#L165) and [fail1-2](https://github.com/ginkgo-project/ginkgo/blob/78d2113dca67afea0d5567dfd511b89bb083e2b9/cuda/solver/gmres_kernels.cu#L188): splitting is failed.
should be dim3(ceildiv(dim_size[0] * stride_next_krylov, default_block_size)), dim3(default_block_size), 0, 0
but get dim3(ceildiv(dim_size[0] * stride_next_krylov)), dim3(default_block_size)), default_block_size, 0

[fail2](https://github.com/ginkgo-project/ginkgo/blob/78d2113dca67afea0d5567dfd511b89bb083e2b9/cuda/solver/gmres_kernels.cu#L245): hipify does not know it is a device call, so hipify does nothing on it.

### CODE REVIEW

**Comment by tcojean** [comment]
@yhmtsai be sure to note down any such problems for a probable future paper

**Comment by tcojean** [comment]
Here is what I tried to make `ginkgo_create_hip_test` work with the CG test. Nothing worked so far. I didn't have much time to explore. If possible @yhmtsai can you take over?
+ I can confirm I don't see constructor anywhere in any of the binaries produced. It is most likely inlined (the heavy use of constexpr in the dim class may have something to do with it). The constructor could potentially be everywhere (hip module, ginkgo core module, and the `cg` test itself should include `dim.hpp`).
+ in `include/ginkgo/core/solver/cg.hpp`, change remove system matrix's `std:move` in case this created a bug, though there should be none: indeed did not change anything.
+ in `include/ginkgo/core/base/exception_helpers.hpp`, change `detail::get_size` to return const reference (hope to not use any copy constructor, though compiler should optimize this), did not seem to work.

**Comment by yhmtsai** [comment]
@tcojean Sure, I will also try to fix it

File: cmake/create_test.cmake [hunk 1]
```diff
@@ -16,6 +16,28 @@ function(ginkgo_create_test test_name)
     add_test(NAME ${REL_BINARY_DIR}/${test_name} COMMAND ${TEST_TARGET_NAME})
 endfunction(ginkgo_create_test)
 
+function(ginkgo_create_test_link_hip test_name)
```
Comment (pratikvn) [inline cmake/create_test.cmake]: I think this is better if named `ginkgo_create_hip_solver_test`.

Comment (yhmtsai) [reply cmake/create_test.cmake]: I would like to use more general one because it might also be used in preconditioner test

Comment (pratikvn) [reply cmake/create_test.cmake]: Then maybe `ginkgo_create_hip_test_special_linkage` is better ?

**Comment by yhmtsai** [comment]
I list the command from three type (gcc/hip/gcc compile hcc compile) output here.
We can go into it in detail at some points to see why the other two are failed.

link
```
== compile:gcc link:hip
/opt/rocm/hip/bin/hipcc_cmake_linker_helper /opt/rocm/hcc    CMakeFiles/hip_test_solver_cg_kernels.dir/cg_kernels.cpp.o -o cg_kernels  -L/usr/lib/gcc/x86_64-linux-gnu/7 ../../../core/libginkgo.a ../../../third_party/gtest/build/googlemock/gtest/./libgtest_main.a ../../../third_party/gtest/build/googlemock/gtest/./libgtest.a ../../../core/device_hooks/libginkgo_omp.a ../../../core/device_hooks/libginkgo_cuda.a ../../../reference/libginkgo_reference.a ../../libginkgo_hip.a /opt/rocm/hip/lib/libhip_hcc.so -L /opt/rocm/lib -Wl,-rpath /opt/rocm/lib -Wl,--whole-archive /opt/rocm/lib/libmcwamp.so.2.9 -Wl,--no-whole-archive -ldl -lm /opt/rocm/lib/libhc_am.so.2.9 /opt/rocm/lib/libhsa-runtime64.so /opt/rocm/hipblas/lib/libhipblas.so.0.1 /opt/rocm/hipsparse/lib/libhipsparse.so.0.1 -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc
== compile:gcc link:gcc
/usr/bin/c++  -O3 -DNDEBUG   CMakeFiles/hip_test_solver_cg_kernels.dir/cg_kernels.cpp.o  -o cg_kernels -Wl,-rpath,/opt/rocm/hip/lib:/opt/rocm/lib:/opt/rocm/hipblas/lib:/opt/rocm/hipsparse/lib ../../../core/libginkgo.a ../../../third_party/gtest/build/googlemock/gtest/./libgtest_main.a ../../../third_party/gtest/build/googlemock/gtest/./libgtest.a ../../../core/device_hooks/libginkgo_omp.a ../../../core/device_hooks/libginkgo_cuda.a ../../../reference/libginkgo_reference.a ../../libginkgo_hip.a /opt/rocm/hip/lib/libhip_hcc.so -L /opt/rocm/lib -Wl,-rpath /opt/rocm/lib -Wl,--whole-archive /opt/rocm/lib/libmcwamp.so.2.9 -Wl,--no-whole-archive -ldl -lm /opt/rocm/lib/libhc_am.so.2.9 /opt/rocm/lib/libhsa-runtime64.so /opt/rocm/hipblas/lib/libhipblas.so.0.1 /opt/rocm/hipsparse/lib/libhipsparse.so.0.1
== compile:hip link:hip
/opt/rocm/hip/bin/hipcc_cmake_linker_helper /opt/rocm/hcc    CMakeFiles/hip_test_solver_cg_kernels.dir/hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o -o cg_kernels  -L/usr/lib/gcc/x86_64-linux-gnu/7 ../../../core/libginkgo.a ../../../third_party/gtest/build/googlemock/gtest/./libgtest_main.a ../../../third_party/gtest/build/googlemock/gtest/./libgtest.a ../../../core/device_hooks/libginkgo_omp.a ../../../core/device_hooks/libginkgo_cuda.a ../../../reference/libginkgo_reference.a ../../libginkgo_hip.a /opt/rocm/hip/lib/libhip_hcc.so -L /opt/rocm/lib -Wl,-rpath /opt/rocm/lib -Wl,--whole-archive /opt/rocm/lib/libmcwamp.so.2.9 -Wl,--no-whole-archive -ldl -lm /opt/rocm/lib/libhc_am.so.2.9 /opt/rocm/lib/libhsa-runtime64.so /opt/rocm/hipblas/lib/libhipblas.so.0.1 /opt/rocm/hipsparse/lib/libhipsparse.so.0.1 -lstdc++ -lm -lgcc_s -lgcc -lc -lgcc_s -lgcc
```
compile 
```
== compile:gcc link:hip
cd /root/ginkgo/build/hip/test/solver && /usr/bin/c++   -I/root/ginkgo/build -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/opt/rocm/hip/include -isystem /root/ginkgo/build/third_party/gtest/src/googletest/include  -O3 -DNDEBUG   -std=gnu++11 -o CMakeFiles/hip_test_solver_cg_kernels.dir/cg_kernels.cpp.o -c /root/ginkgo/hip/test/solver/cg_kernels.cpp
== compile:gcc link:gcc
cd /root/ginkgo/build/hip/test/solver && /usr/bin/c++   -I/root/ginkgo/build -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/opt/rocm/hip/include -isystem /root/ginkgo/build/third_party/gtest/src/googletest/include  -O3 -DNDEBUG   -std=gnu++11 -o CMakeFiles/hip_test_solver_cg_kernels.dir/cg_kernels.cpp.o -c /root/ginkgo/hip/test/solver/cg_kernels.cpp
== compile:hip link:hip
Building HIPCC object hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir/hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o
cd /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir && /usr/local/bin/cmake -E make_directory /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//.
cd /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir && /usr/local/bin/cmake -D verbose:BOOL=1 -D build_configuration:STRING=RELEASE -D generated_file:STRING=/root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//./hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o -P /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.cmake
-- Removing /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//./hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o
/usr/local/bin/cmake -E remove /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//./hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o
-- Generating dependency file: /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.pre
/opt/rocm/hip/bin/hipcc -M /root/ginkgo/hip/test/solver/cg_kernels.hip.cpp -o /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.pre -std=c++11 "-I/root/ginkgo/build -I/opt/rocm/include -I/opt/rocm/hipblas/include -I/opt/rocm/hipsparse/include -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/opt/rocm/hip/include -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/root/ginkgo/build/third_party/gtest/src/googletest/include -I/root/ginkgo/build/third_party/gtest/src/googletest/include"
-- Generating temporary cmake readable file: /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.tmp
/usr/local/bin/cmake -D input_file:FILEPATH=/root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.pre -D output_file:FILEPATH=/root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.tmp -D verbose=1 -P /opt/rocm/hip/cmake/FindHIP/run_make2cmake.cmake
-- Copy if different /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.tmp to /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend
/usr/local/bin/cmake -E copy_if_different /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.tmp /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend
-- Removing /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.tmp and /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.pre
/usr/local/bin/cmake -E remove /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.tmp /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o.depend.pre
-- Generating /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//./hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o
/opt/rocm/hip/bin/hipcc -c /root/ginkgo/hip/test/solver/cg_kernels.hip.cpp -o /root/ginkgo/build/hip/test/solver/CMakeFiles/hip_test_solver_cg_kernels.dir//./hip_test_solver_cg_kernels_generated_cg_kernels.hip.cpp.o -std=c++11 "-I/root/ginkgo/build -I/opt/rocm/include -I/opt/rocm/hipblas/include -I/opt/rocm/hipsparse/include -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/opt/rocm/hip/include -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/root/ginkgo/build/include -I/root/ginkgo/include -I/root/ginkgo -I/root/ginkgo/build/third_party/gtest/src/googletest/include -I/root/ginkgo/build/third_party/gtest/src/googletest/include"
```

**Comment by tcojean** [comment]
@yhmtsai so what works for now is building with gcc and linking with HIP for these tests?

**Comment by yhmtsai** [comment]
@tcojean Yes, building with gcc and linking with hcc works, but I do not know how it works.
Only use this way when build ginkgo static library and the platform is hcc. Others still use gcc to compile and link.

**Pull request merged**